aiidalab/environment.py

```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+"""Manage Python environments and Jupyter kernels for apps."""
+import re
+import venv
+import shutil
+from hashlib import sha1
+from pathlib import Path
+from subprocess import run
+from urllib.parse import quote_plus
+
+from .config import AIIDALAB_APPS
+
+
+def _valid_jupyter_kernel_name(name):
+    """Check whether the given name is a valid Jupyter kernel name."""
+    return re.fullmatch(r'[a-z0-9\-\_.]+', name)
+
+
+class AppEnvironmentError(RuntimeError):
+    """Indicates issue relating to the AppEnvironment installation."""
+
+
+class AppEnvironment:
+    """Manage Python environment and Jupyter kernel for AiiDA lab app.
+
+    Arguments:
+
+        app_name (str):
+            Name of the AiiDA lab app.
+    """
+
+    def __init__(self, app_name):
+        self._app_name = app_name
+
+    @property
+    def kernel_name(self):
+        """Generate unique name for this kernel based on the app name.
+
+        The kernel name is guaranteed to be a valid Jupyter kernel name
+        and unique with respect to the app name so to avoid kernel name
+        collisions between different apps.
+        """
+        # The unique_name is a hash value based on the app_name which is
+        # guaranteed to be truly unique with respect to the app name.
+        # The current schema version (1) is explicitly encoded to simplify
+        # potential future migrations.
+        unique_name = sha1(f'1/{self._app_name}'.encode()).hexdigest()
+
+        # The human-readable name is a name based on the app name that
+        # constitutes a valid Jupyter kernel name:
+        human_readable_name = quote_plus(self._app_name.replace('@', '.')).replace('%', '-')
+        kernel_name = f'{human_readable_name:.32}-{unique_name:.8}'
+
+        # Return only the unique_name in case that we failed to
+        # construct a valid kernel name:
+        if not _valid_jupyter_kernel_name(kernel_name):
+            return unique_name
+
+        return kernel_name
+
+    @property
+    def prefix(self):
+        return Path(AIIDALAB_APPS).joinpath(self._app_name, '.venv')
+
+    @property
+    def executable(self):
+        return self.prefix.joinpath('bin', 'python')
+
+    @property
+    def jupyter_kernel_path(self):
+        return Path.home().joinpath('.local', 'share', 'jupyter', 'kernels', self.kernel_name)
```

Comment:
Maybe let's have a look if we can determine it programmatically

Reply:
Found a solution in combination with `jupyter_core.paths.jupyter_data_dir()`.